### PR TITLE
I23 nonlinear programming

### PR DESCRIPTION
**Pull request by neginlashkari**
Summary of code improvements:

Changes have been done on main codes in order to:

1- Combine all figures in one place to run the animation
2- Save a .GIF file from final animation in the current path
3- Change the vehicle model to a two-wheeled car
4- Add velocity constraint of robot
5- Use Bezier curves to generate random desired trajectory
6- Display if the problem has feasible solution using exit flag

### CODE REVIEW

**Comment by yshiyi** [comment]
fig2
I see you have changed the vehicle model to a two-wheeled car. Then there is a problem. You may see it form the figure above. The edge of the car and the edge of the obstacle are overlapping. Maybe add a constraint of the minimum distance between the center of the car and the edge of the obstacle.

**Comment by yshiyi** [comment]
fig1
fig3
I've got two fail results which are shown in the above figures. I guess the reason is that the "fmincon" function doesn't always result a optimal solution. I am trying to find out what the exact reason for this failure is and will let you know if I've got anything.

**Comment by yshiyi** [comment]
fig4
I've also got this error message. Because "fmincon" fails again, I guess.

**Comment by yshiyi** [comment]
In addition, please consider commenting on that each of state represents for.
It is a good idea to use Bezier curves to generate the desired trajectory. Overall, I think it is a good work. It looks clean and neat. In the end, please check the cases I list above.

File: matlab_simulation/code v1.0/ME597-8-NLP/Trajectory_bazier.m [hunk 1]
```diff
@@ -1,5 +1,20 @@
+%% This file develops a desired trajectory using Bazier Curves.
```
Comment (stevenwaslander) [inline matlab_simulation/code v1.0/ME597-8-NLP/Trajectory_bazier.m:1]: Bezier, not Bazier, switch throughout.

File: matlab_simulation/code v1.0/ME597-8-NLP/Bz_fun.m [hunk 2]
```diff
@@ -0,0 +1,9 @@
+function Sp = Bz_fun(g)
+%BZ_FUN
+%    SP = BZ_FUN(G)
+
+%    This function was generated by the Symbolic Math Toolbox version 6.1.
```
Comment (stevenwaslander) [inline matlab_simulation/code v1.0/ME597-8-NLP/Bz_fun.m:5]: Why is this curve hard coded? If you want to use Bezier curves, make it possible to generate them on the fly for each problem instance.  I'm fine with just using sine or cosine trajectories, though, as Bezier would be another thing to explain in class.

**Comment by stevenwaslander** [comment]
It would be best to remove the animation, but to place an example code to generate the animation in the 01-examples-lecture folder.  Move the function that does the optimization to the planning folder, and any common tools like bezier curve generation to the utilities folder.

You'll also need to resolve the merge conflict locally (figure out what conflicts, and then fix it and commit the fix locally, then push the commit to your branch on the remote (github)).

**Comment by stevenwaslander** [comment]
Please add comments throughout the code explaining what you are doing, especially when it comes to indexes in this problem.

**Comment by stevenwaslander** [comment]
Windows needs lower-case fmincon, upper case fails.

When I switch, I get the following indexing error every time.
Subscripted assignment dimension mismatch.

Error in constraints (line 28)
C(ind_end+1,1)=x(4)-vd_cnst;

Error in @(x)constraints(x)

Error in fmincon (line 632)
        [ctmp,ceqtmp] = feval(confcn{3},X,varargin{:});

Error in nonlinearplan (line 105)
[X,FVAL,EXITFLAG,OUTPUT,LAMBDA] = fmincon(@(x)
cost(x),x0,A,B,Aeq,Beq,LB,UB,@(x) constraints(x), options);

Caused by:
    Failure in initial user-supplied nonlinear constraint function
    evaluation. FMINCON cannot continue.

File: matlab_simulation/code v1.0/ME597-8-NLP/Trajectory_bazier.m [hunk 3]
```diff
@@ -1,5 +1,20 @@
...
+%% P is a set of points including start point, bazier control point and target.
+
 function xdT=Trajectory_bazier(xd_start,xd_end,dt,TTot)
 xd_midpoint=(xd_end(1)-xd_start(1)).*(rand(1,2));
 P=[xd_start;xd_start+xd_midpoint;xd_end];
-BezierCurve(P); %generates Bz_fun
+
+syms g
+Np = size(P, 1); 
+
```
Comment (stevenwaslander) [inline matlab_simulation/code v1.0/ME597-8-NLP/Trajectory_bazier.m:10]: What is all this? Explain what you are doing in comments.  Add references or links to wikipedia if not covered in the course notes.

File: matlab_simulation/code v1.0/ME597-8-NLP/Trajectory_bazier.m [hunk 4]
```diff
@@ -1,5 +1,20 @@
...
+for i = 1:Np
+   Bp(:,i) = nchoosek(Np,i-1).*(g.^(i-1)).*((1-g).^(Np-i+1)); %Bp is the Bernstein polynomial value
+end
+
+Bp1= (nchoosek(Np,Np).*(g.^Np)).';
+
+Sp= simplify(Bp*P + Bp1*P(Np,:));
+Bz_fun=matlabFunction(Sp);
```
Comment (stevenwaslander) [inline matlab_simulation/code v1.0/ME597-8-NLP/Trajectory_bazier.m:18]: What does this line do? matlabFunction?

File: matlab_simulation/code v1.0/ME597-8-NLP/circle.m [hunk 5]
```diff
@@ -1,5 +1,7 @@
+% This function draws cicles given the position of center and the radius
```
Comment (stevenwaslander) [inline matlab_simulation/code v1.0/ME597-8-NLP/circle.m:1]: Should be in utilities with the other plotting functions.

File: matlab_simulation/code v1.0/ME597-8-NLP/constraints.m [hunk 6]
```diff
@@ -22,3 +22,7 @@
         end
     end
 end
+
+% Velocity constraints
```
Comment (stevenwaslander) [inline matlab_simulation/code v1.0/ME597-8-NLP/constraints.m]: These are variable bound constraints, I believe, and should be applied at every time step.

File: matlab_simulation/code v1.0/ME597-8-NLP/nonlinearrhc.m [hunk 7]
```diff
@@ -34,12 +39,14 @@
 % Receding Horizon
 T = 5;
 
-% Constraints
+% Example #4 Constraints (velocity constraint is added)
```
Comment (stevenwaslander) [inline matlab_simulation/code v1.0/ME597-8-NLP/nonlinearrhc.m]: Would be nicer if each example was its own file in the examples-lectures folder

File: matlab_simulation/code v1.0/ME597-8-NLP/nonlinearrhc.m [hunk 8]
```diff
@@ -75,59 +82,52 @@
...
-%     plot(1:T,x)
-%     plot(1:T,y)
-%     plot(1:T,th)
-%     plot(1:T,v)
-%     plot(1:T,w)
-
     figure(ii); clf; hold on;
     plot(x,y,'bx-');
+    % draw vehicle as a two-wheeled car
+    drawcar(x(end),y(end),th(end),.15,ii)
```
Comment (stevenwaslander) [inline matlab_simulation/code v1.0/ME597-8-NLP/nonlinearrhc.m]: The vehicle should be at the start of the trajectory, not the end.  The plan is into the future, and the vehicle executes only the first step, then replans.

File: matlab_simulation/code v1.0/ME597-8-NLP/nonlinearrhc.m [hunk 9]
```diff
@@ -75,59 +82,52 @@
...
             circle(ii, obs(j,:), radius(j));
         end
         axis equal
     end
-    %drawnow();
     F(i) = getframe;
     
-    % saving final animation file as a .gif file
+    
+    % saving animation file in .gif file
```
Comment (stevenwaslander) [inline matlab_simulation/code v1.0/ME597-8-NLP/nonlinearrhc.m:130]: Animating using a .gif is cool, but can you make it a function in utilities that you call?

**Comment by stevenwaslander** [comment]
Have modified code and merged directly into master.  This pull request pushes changes the wrong way.

**Pull request closed without merging**